src/sbvr-api/uri-parser.ts

```diff
@@ -0,0 +1,225 @@
+}
+
+const memoizedOdata2AbstractSQL = (() => {
+	const memoizedOdata2AbstractSQL = memoize(
+		(vocabulary: string, odataQuery: ODataQuery, method: SupportedMethod, bodyKeys: string[]) => {
+			try {
+				return odata2AbstractSQL[vocabulary].match(odataQuery, 'Process', [method, bodyKeys])
+			} catch (e) {
+				console.error('Failed to translate url: ', JSON.stringify(odataQuery, null, '\t'), method, e, e.stack)
+				throw new TranslationError('Failed to translate url')
+			}
+		},
+		{ normalizer: JSON.stringify }
+	)
+	return (vocabulary: string, odataQuery: ODataQuery, method: SupportedMethod, body: {}) => {
+		// Sort the body keys to improve cache hits
+		const { tree, extraBodyVars } = memoizedOdata2AbstractSQL(vocabulary, odataQuery, method, _.keys(body).sort())
+		_.assign(body, extraBodyVars)
+		return tree
+	}
+})()
+
+export const metadataEndpoints = [ '$metadata', '$serviceroot' ]
+
+const notBadRequestOrParsingError = (e: Error) => {
+	return !((e instanceof BadRequestError) || (e instanceof ParsingError))
+}
+
+export const parseOData = (b: UnparsedRequest) => {
+	return Promise.try<ODataRequest | ODataRequest[]>(() => {
+		if (b._isChangeSet && b.changeSet != null) {
+			const env = new Map<ODataRequest['id'], ODataRequest>()
+			// We sort the CS set once, we must assure that requests which reference
+			// other requests in the changeset are placed last. Once they are sorted
+			// Map will guarantee retrival of results in insertion order
+			const sortedCS = _.sortBy(b.changeSet, (el) => el.url[0] !== '/')
+			return Promise.reduce(sortedCS, parseODataChangeset, env)
+			.then((env) => Array.from(env.values()) as ODataRequest[])
+		} else {
+			const { url, apiRoot } = splitApiRoot(b.url)
+			const odata = odataParser.matchAll(url, 'Process')
+
+			return {
+				method: b.method as SupportedMethod,
+				vocabulary: apiRoot,
+				resourceName: odata.tree.resource,
+				odataBinds: odata.binds,
+				odataQuery: odata.tree,
+				values: b.data,
+				custom: {},
+				_defer: false,
+			}
+		}
+	}).catch(SyntaxError, () => {
+		throw new BadRequestError(`Malformed url: '${b.url}'`)
+	}).catch(notBadRequestOrParsingError, (e) => {
+		console.error('Failed to parse url: ', b.method, b.url, e, e.stack)
+		throw new ParsingError(`Failed to parse url: '${b.url}'`)
+	})
+}
+
+const parseODataChangeset = (env: Map<ODataRequest['id'], ODataRequest>, b: UnparsedRequest) => {
+	const contentId: ODataRequest['id'] = mustExtractHeader(b, 'content-id')
+
+	if (env.has(contentId)) {
+		throw new BadRequestError('Content-Id must be unique inside a changeset')
+	}
+
+	let defer: boolean
+	let odata
+	let apiRoot: string
+	let url
+
+	if (b.url[0] === '/') {
+		({ url, apiRoot } = splitApiRoot(b.url))
+		odata = odataParser.matchAll(url, 'Process')
+		defer = false
+	} else {
+		url = b.url
+		odata = odataParser.matchAll(url, 'Process')
+		const { bind } = odata.tree.resource
+		const [ , id ] = odata.binds[bind]
```

Comment:
Destructuring is Good, but why not write `const id = odata.binds[bind][1]` instead?

Reply:
Because this matches how it was in the coffeescript version ( https://github.com/resin-io/pinejs/pull/82/files#diff-c52cf3760792616d7d710ad38b04787dL86 ) and in general I try to avoid changes beyond what is necessary for typescript to be happy (to minimise the chance of introducing an unintended issue)